src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs

```diff
@@ -30,20 +35,66 @@ public static void PopulateDefaults(PageRouteModel model, string routeTemplate)
                 // force outgoing routes to match to the path sans /Index.
                 selectorModel.AttributeRouteModel.SuppressLinkGeneration = true;
 
-                var parentDirectoryPath = model.ViewEnginePath;
-                var index = parentDirectoryPath.LastIndexOf('/');
-                if (index == -1)
-                {
-                    parentDirectoryPath = string.Empty;
-                }
-                else
-                {
-                    parentDirectoryPath = parentDirectoryPath.Substring(0, index);
-                }
+                var index = pageRoute.LastIndexOf('/');
+                var parentDirectoryPath = index == -1 ?
+                    string.Empty :
+                    pageRoute.Substring(0, index);
                 model.Selectors.Add(CreateSelectorModel(parentDirectoryPath, routeTemplate));
             }
         }
 
+        public static bool TryParseAreaPath(
+            RazorPagesOptions razorPagesOptions,
+            string path,
+            ILogger logger,
+            out (string areaName, string viewEnginePath, string pageRoute) result)
+        {
+            // path = "/Products/Pages/Manage/Home.cshtml"
+            // Result = ("Products", "/Manage/Home", "/Products/Manage/Home")
+
+            result = default;
+            Debug.Assert(path.StartsWith("/", StringComparison.Ordinal));
+
+            var areaEndIndex = path.IndexOf('/', startIndex: 1);
+            if (areaEndIndex == -1 || areaEndIndex == path.Length)
+            {
+                logger?.UnsupportedAreaPath(razorPagesOptions, path);
+                return false;
+            }
+
+            // Normalize the pages root directory so that it has a 
+            var normalizedPagesRootDirectory = razorPagesOptions.RootDirectory.TrimStart('/');
+            if (!normalizedPagesRootDirectory.EndsWith("/", StringComparison.Ordinal))
+            {
+                normalizedPagesRootDirectory += "/";
+            }
+
+            if (string.Compare(path, areaEndIndex + 1, normalizedPagesRootDirectory, 0, normalizedPagesRootDirectory.Length, StringComparison.OrdinalIgnoreCase) != 0)
+            {
+                logger?.UnsupportedAreaPath(razorPagesOptions, path);
+                return false;
+            }
+
+            var areaName = path.Substring(1, areaEndIndex - 1);
+
+            var pagePathIndex = areaEndIndex + normalizedPagesRootDirectory.Length;
+            var pageExtensionIndex = path.LastIndexOf('.');
+            var pageExtensionLength = pageExtensionIndex == -1 || pageExtensionIndex < pagePathIndex ?
+                0 :
+                path.Length - pageExtensionIndex;
+
+            // PageName
+            var pageName = path.Substring(pagePathIndex, path.Length - pagePathIndex - pageExtensionLength);
+
+            var builder = new InplaceStringBuilder(areaEndIndex + pageName.Length);
+            builder.Append(path, 0, areaEndIndex);
```

Comment:
If you're starting at zero isn't this just a string.Concat?

Reply:
Doh.

Reply:
It used to be a bit messier which is why I thought I missed something obvious AFAIK, there isn't a way to concat a substring without calling `.Substring` that I wanted to avoid here.